### PR TITLE
`/transaction/cost` fix endpoint

### PR DESCRIPTION
**Pull request by miiu96**
## Reasoning behind the pull request
- The endpoint `/transaction/cost/` needs to be fixed in order to accurately compute the cost of a provided transaction in gas units. Currently, there are issues with the computation process, resulting in incorrect cost values. The fix aims to address these issues and ensure that the gas cost of the transaction is calculated accurately.
## Proposed changes
- In this PR the response of the endpoint `/transaction/cost/` is extended to include the generated logs from the simulation of the transaction. Previously, the endpoint only provided the computed cost in gas units. However, to enhance transparency and debugging capabilities, it was deemed necessary to include the simulation logs as part of the response.

## Testing procedure
- Call this endpoint with different types of transactions like move balance transactions, ESDTs transactions, and smart contract calls and check if the cost from the response is the same with the `gasUsed` field from the transactions that are indexed in the Elasticsearch database.
## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: factory/processing/blockProcessorCreator.go [hunk 1]
```diff
@@ -1116,6 +1116,11 @@ func (pcf *processComponentsFactory) createMetaTxSimulatorProcessor(
 		return nil, err
 	}
 
+	err = builtInFuncFactory.SetPayableHandler(vmFactory.BlockChainHookImpl())
```
Comment (iulianpascalau) [inline factory/processing/blockProcessorCreator.go:1119]: I think this is already fixed in the PR that changes the tx simulation construction. You should check at merge time when resolving the (future) conflict :D

File: go.mod [hunk 2]
```diff
@@ -13,7 +13,7 @@ require (
 	github.com/google/gops v0.3.18
 	github.com/gorilla/websocket v1.5.0
 	github.com/mitchellh/mapstructure v1.5.0
-	github.com/multiversx/mx-chain-core-go v1.2.3
+	github.com/multiversx/mx-chain-core-go v1.2.4-0.20230517101712-1ac901fd7b60
```
Comment (iulianpascalau) [inline go.mod:16]: proper release

File: node/external/nodeApiResolver.go [hunk 3]
```diff
@@ -131,6 +132,10 @@ func (nar *nodeApiResolver) ComputeTransactionGasLimit(tx *transaction.Transacti
 	return nar.txCostHandler.ComputeTransactionGasLimit(tx)
 }
 
+func (nar *nodeApiResolver) SimulateTransactionExecution(tx *transaction.Transaction) (*txSimData.SimulationResults, error) {
```
Comment (iulianpascalau) [inline node/external/nodeApiResolver.go:135]: missing comment, also, why do we have a new function for the tx simulation? I remember we already have 3 of them. Maybe clean them up?

File: node/mock/transactionCostEstimatorMock.go [hunk 4]
```diff
@@ -15,6 +19,15 @@ func (tcem *TransactionCostEstimatorMock) ComputeTransactionGasLimit(tx *transac
 	return &transaction.CostResponse{}, nil
 }
 
+// SimulateTransactionExecution ---
```
Comment (iulianpascalau) [inline node/mock/transactionCostEstimatorMock.go:22]: only one `-` ?

File: process/transaction/transactionCostEstimator.go [hunk 5]
```diff
@@ -74,10 +74,24 @@ func NewTransactionCostEstimator(
 	return tce, nil
 }
 
+// SimulateTransactionExecution will simulate a transaction's execution and will return the results
+func (tce *transactionCostEstimator) SimulateTransactionExecution(tx *transaction.Transaction) (*txSimData.SimulationResults, error) {
+	tce.mutExecution.Lock()
```
Comment (iulianpascalau) [inline process/transaction/transactionCostEstimator.go:79]: :-1: this ensures all API calls will execute sequentially. The results should be correct in this way.

File: process/txsimulator/data/data.go [hunk 6]
```diff
@@ -7,10 +7,6 @@ import (
 
 // SimulationResults is the data transfer object which will hold results for simulation a transaction's execution
 type SimulationResults struct {
```
Comment (iulianpascalau) [inline process/txsimulator/data/data.go:9]: SimulationResultsWithVMOutput ? In order to distinguish between `transaction.SimulationResults` and `data.SimulationResults`

File: process/txsimulator/interface.go [hunk 7]
```diff
@@ -11,3 +12,8 @@ type TransactionProcessor interface {
 	VerifyTransaction(transaction *transaction.Transaction) error
 	IsInterfaceNil() bool
 }
+
+// DataFieldParser defines what a data field parser should be able to do
+type DataFieldParser interface {
+	Parse(dataField []byte, sender, receiver []byte, numOfShards uint32) *datafield.ResponseParseData
```
Comment (iulianpascalau) [inline process/txsimulator/interface.go:18]: IsInterfaceNil ?

File: process/txsimulator/txSimulator.go [hunk 8]
```diff
@@ -124,7 +165,7 @@ func (ts *transactionSimulator) getVMOutputOfTx(tx *transaction.Transaction) (*v
 	defer ts.vmOutputCacher.Remove(txHash)
 
 	vmOutputI, ok := ts.vmOutputCacher.Get(txHash)
-	if !ok {
+	if !ok || check.IfNilReflect(vmOutputI) {
```
Comment (iulianpascalau) [inline process/txsimulator/txSimulator.go:168]: what should `vmOutoutI` mean? :D

File: process/txsimulator/wrappedAccountsDB.go [hunk 9]
```diff
@@ -11,6 +12,8 @@ import (
 
 // readOnlyAccountsDB is a wrapper over an accounts db which works read-only. write operation are disabled
 type readOnlyAccountsDB struct {
```
Comment (iulianpascalau) [inline process/txsimulator/wrappedAccountsDB.go:14]: let's rename this to simulationAccountsDB (+ the file) as this no longer is a readonly instance

File: process/txsimulator/wrappedAccountsDB.go [hunk 10]
```diff
@@ -11,6 +12,8 @@ import (
...
+	mutex            sync.RWMutex
+	cashedAccounts   map[string]vmcommon.AccountHandler
```
Comment (iulianpascalau) [inline process/txsimulator/wrappedAccountsDB.go:16]: cachedAccounts :D

File: process/txsimulator/wrappedAccountsDB.go [hunk 11]
```diff
@@ -148,3 +178,10 @@ func (r *readOnlyAccountsDB) Close() error {
 func (r *readOnlyAccountsDB) IsInterfaceNil() bool {
 	return r == nil
 }
+
+// CleanCache -
```
Comment (iulianpascalau) [inline process/txsimulator/wrappedAccountsDB.go:182]: meaningful comment?

File: testscommon/state/accountsAdapterStub.go [hunk 12]
```diff
@@ -42,6 +42,10 @@ type AccountsStub struct {
 	StartSnapshotIfNeededCalled   func() error
 }
 
+// CleanCache --
```
Comment (iulianpascalau) [inline testscommon/state/accountsAdapterStub.go:45]: one `-` ?

File: factory/processing/blockProcessorCreator.go (hunk 1 above)
Comment (miiu96) [reply factory/processing/blockProcessorCreator.go]: fixed

File: go.mod (hunk 2 above)
Comment (miiu96) [reply go.mod:16]: will add after the review is done

File: node/mock/transactionCostEstimatorMock.go (hunk 4 above)
Comment (miiu96) [reply node/mock/transactionCostEstimatorMock.go:22]: done

File: node/external/nodeApiResolver.go (hunk 3 above)
Comment (miiu96) [reply node/external/nodeApiResolver.go:135]: added comment

File: process/txsimulator/data/data.go (hunk 6 above)
Comment (miiu96) [reply process/txsimulator/data/data.go:9]: done

File: process/txsimulator/interface.go (hunk 7 above)
Comment (miiu96) [reply process/txsimulator/interface.go:18]: the data field parser from the `mx-chain-vm-common-go` does not implement this interface.

File: process/txsimulator/txSimulator.go (hunk 8 above)
Comment (miiu96) [reply process/txsimulator/txSimulator.go:168]: renamed

File: process/txsimulator/wrappedAccountsDB.go (hunk 10 above)
Comment (miiu96) [reply process/txsimulator/wrappedAccountsDB.go:16]: done

File: process/txsimulator/wrappedAccountsDB.go (hunk 9 above)
Comment (miiu96) [reply process/txsimulator/wrappedAccountsDB.go:14]: done.

File: process/txsimulator/wrappedAccountsDB.go (hunk 11 above)
Comment (miiu96) [reply process/txsimulator/wrappedAccountsDB.go:182]: done

File: testscommon/state/accountsAdapterStub.go (hunk 12 above)
Comment (miiu96) [reply testscommon/state/accountsAdapterStub.go:45]: done

File: factory/interface.go [hunk 13]
```diff
@@ -256,9 +256,10 @@ type NetworkComponentsHandler interface {
 	NetworkComponentsHolder
 }
 
-// TransactionSimulatorProcessor defines the actions which a transaction simulator processor has to implement
-type TransactionSimulatorProcessor interface {
-	ProcessTx(tx *transaction.Transaction) (*txSimData.SimulationResults, error)
+// TransactionCostSimulator defines the transaction cost simulator actions
+type TransactionCostSimulator interface {
```
Comment (iulianpascalau) [inline factory/interface.go:260]: The main implementation for this is called TransactionCostEstimator, why not renaming the interface to `TransactionCostEstimator` ?

File: integrationTests/testProcessorNodeWithTestWebServer.go [hunk 14]
```diff
@@ -176,16 +175,16 @@ func createFacadeComponents(tpn *TestProcessorNode) nodeFacade.ApiResolver {
...
 	log.LogIfError(err)
-	txCostHandler, err := transaction.NewTransactionCostEstimator(
-		txTypeHandler,
-		tpn.EconomicsData,
-		txSimulator,
-		wrappedAccounts,
-		tpn.ShardCoordinator,
-		tpn.EnableEpochsHandler,
-	)
+	txCostHandler, err := txsimulator.NewTransactionCostEstimator(txsimulator.ArgsTransactionCostSimulator{
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNodeWithTestWebServer.go:180]: could have been an argument variable

File: integrationTests/vm/testInitializer.go [hunk 15]
```diff
@@ -959,14 +959,14 @@ func CreateTxProcessorWithOneSCExecutorWithVMs(
...
 
-	txCostEstimator, err := transaction.NewTransactionCostEstimator(
-		txTypeHandler,
-		economicsData,
-		txSimulator,
-		readOnlyAccountsDB,
-		shardCoordinator,
-		argsNewSCProcessor.EnableEpochsHandler,
-	)
+	txCostEstimator, err := txsimulator.NewTransactionCostEstimator(txsimulator.ArgsTransactionCostSimulator{
```
Comment (iulianpascalau) [inline integrationTests/vm/testInitializer.go:962]: could have been an argument variable

File: process/txsimulator/simulationAccountsDB_test.go [hunk 16]
```diff
@@ -17,15 +17,15 @@ import (
 func TestNewReadOnlyAccountsDB_NilOriginalAccountsDBShouldErr(t *testing.T) {
 	t.Parallel()
 
-	roAccDb, err := NewReadOnlyAccountsDB(nil)
+	roAccDb, err := NewSimulationAccountsDB(nil)
```
Comment (iulianpascalau) [inline process/txsimulator/simulationAccountsDB_test.go:20]: rename the variable to accountsDB ? (ro came from read-only)
Valid for all tests

File: process/txsimulator/transactionCostEstimator.go [hunk 17]
```diff
@@ -24,6 +23,15 @@ const dummySignature = "01010101"
 const gasRemainedSplitString = "gas remained = "
 const gasUsedSlitString = "gas used = "
 
+type ArgsTransactionCostSimulator struct {
```
Comment (iulianpascalau) [inline process/txsimulator/transactionCostEstimator.go:26]: missing comments

File: integrationTests/vm/testInitializer.go [hunk 18]
```diff
@@ -875,7 +875,7 @@ func CreateTxProcessorWithOneSCExecutorWithVMs(
 	}
 
 	// create transaction simulator
-	readOnlyAccountsDB, err := txsimulator.NewReadOnlyAccountsDB(accnts)
+	readOnlyAccountsDB, err := txsimulator.NewSimulationAccountsDB(accnts)
```
Comment (bogdan-rosianu) [inline integrationTests/vm/testInitializer.go:878]: rename var to `simulationAccountsDB`

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: process/txsimulator/simulationAccountsDB.go [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+// GetStackDebugFirstEntry -
+func (r *simulationAccountsDB) GetStackDebugFirstEntry() []byte {
+	return nil
+}
+
+// Close will handle the closing of the underlying components
+func (r *simulationAccountsDB) Close() error {
+	return r.originalAccounts.Close()
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/txsimulator/simulationAccountsDB.go:174]: good find

File: factory/interface.go (hunk 13 above)
Comment (miiu96) [reply factory/interface.go:260]: refactored

File: integrationTests/testProcessorNodeWithTestWebServer.go (hunk 14 above)
Comment (miiu96) [reply integrationTests/testProcessorNodeWithTestWebServer.go:180]: done

File: integrationTests/vm/testInitializer.go (hunk 15 above)
Comment (miiu96) [reply integrationTests/vm/testInitializer.go:962]: done

File: process/txsimulator/simulationAccountsDB_test.go (hunk 16 above)
Comment (miiu96) [reply process/txsimulator/simulationAccountsDB_test.go:20]: done

File: process/txsimulator/transactionCostEstimator.go (hunk 17 above)
Comment (miiu96) [reply process/txsimulator/transactionCostEstimator.go:26]: done

File: factory/processing/txSimulatorProcessComponents.go [hunk 20]
```diff
@@ -55,15 +55,27 @@ func (pcf *processComponentsFactory) createTxSimulatorProcessor() (factory.Trans
 	txSimulatorProcessorArgs.Marshalizer = pcf.coreData.InternalMarshalizer()
 
 	txSimulator, err := txsimulator.NewTransactionSimulator(txSimulatorProcessorArgs)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	txCostSimulator, err := txsimulator.NewTransactionCostEstimator(txsimulator.ArgsTransactionCostSimulator{
```
[the remark this answers is not in the document]
Comment (miiu96) [reply factory/processing/txSimulatorProcessComponents.go:62]: refactored

File: integrationTests/vm/testInitializer.go (hunk 18 above)
Comment (miiu96) [reply integrationTests/vm/testInitializer.go:878]: done

File: process/txsimulator/simulationAccountsDB.go [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+// LoadAccount will call the original accounts' function with the same name
+func (r *simulationAccountsDB) LoadAccount(address []byte) (vmcommon.AccountHandler, error) {
+	r.mutex.RLock()
+	cachedAccount, ok := r.cachedAccounts[string(address)]
+	r.mutex.RUnlock()
+	if ok {
+		return cachedAccount, nil
+	}
+
+	return r.originalAccounts.LoadAccount(address)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/txsimulator/simulationAccountsDB.go:74]: added

File: process/txsimulator/simulationAccountsDB.go [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+// GetExistingAccount will call the original accounts' function with the same name
+func (r *simulationAccountsDB) GetExistingAccount(address []byte) (vmcommon.AccountHandler, error) {
+	r.mutex.RLock()
+	cachedAccount, ok := r.cachedAccounts[string(address)]
+	r.mutex.RUnlock()
+	if ok {
+		return cachedAccount, nil
+	}
+
+	return r.originalAccounts.GetExistingAccount(address)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/txsimulator/simulationAccountsDB.go:57]: done.

File: process/txsimulator/simulationAccountsDB.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply process/txsimulator/simulationAccountsDB.go:174]: refactored.

File: process/txsimulator/txSimulator.go [hunk 23]
```diff
@@ -68,6 +76,14 @@ func NewTransactionSimulator(args ArgsTxSimulator) (*transactionSimulator, error
 		return nil, ErrNilHasher
 	}
 
+	dataFieldParser, err := datafield.NewOperationDataFieldParser(&datafield.ArgsOperationDataFieldParser{
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/txsimulator/txSimulator.go:79]: done.

File: factory/mock/processComponentsStub.go [hunk 24]
```diff
@@ -187,7 +187,7 @@ func (pcm *ProcessComponentsMock) FallbackHeaderValidator() process.FallbackHead
 }
 
 // TransactionCostSimulator -
-func (pcm *ProcessComponentsMock) TransactionCostSimulator() factory.TransactionCostSimulator {
+func (pcm *ProcessComponentsMock) APITransactionEvaluator() factory.TransactionEvaluator {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply factory/mock/processComponentsStub.go:190]: fixed

File: factory/processing/export_test.go [hunk 25]
```diff
@@ -47,6 +47,6 @@ func (pcf *processComponentsFactory) NewBlockProcessor(
 }
 
 // CreateTxSimulatorProcessor -
-func (pcf *processComponentsFactory) CreateTxSimulatorProcessor() (factory.TransactionCostSimulator, process.VirtualMachinesContainerFactory, error) {
-	return pcf.createTxCostSimulator()
+func (pcf *processComponentsFactory) CreateTxSimulatorProcessor() (factory.TransactionEvaluator, process.VirtualMachinesContainerFactory, error) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply factory/processing/export_test.go]: done

File: factory/mock/processComponentsStub.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply factory/mock/processComponentsStub.go:190]: pcm.TxCostSimulator - still not done

File: factory/processing/txSimulatorProcessComponents_test.go [hunk 26]
```diff
@@ -26,26 +26,26 @@ func TestManagedProcessComponents_createTxSimulatorProcessor(t *testing.T) {
...
-		assert.False(t, check.IfNil(txSimulator))
+		assert.False(t, check.IfNil(apiTransactionEvaluator))
 		assert.False(t, check.IfNil(vmContainerFactory))
 	})
 	t.Run("should work for metachain", func(t *testing.T) {
 		processArgs := components.GetProcessComponentsFactoryArgs(shardCoordinatorForMetachain)
 		pcf, _ := processing.NewProcessComponentsFactory(processArgs)
 
-		txSimulator, vmContainerFactory, err := pcf.CreateTxSimulatorProcessor()
+		txSimulator, vmContainerFactory, err := pcf.CreateAPITransactionEvaluator()
```
Comment (bogdan-rosianu) [inline factory/processing/txSimulatorProcessComponents_test.go:48]: txSimulator -> txEvaluator

Comment (miiu96) [reply factory/processing/txSimulatorProcessComponents_test.go:48]: done.

File: factory/mock/processComponentsStub.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply factory/mock/processComponentsStub.go:190]: done

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — DISMISSED** [review]
System test passed. 
@@                            Log scanner                                  @@
##                            fix-tx-cost-api-route                        ##
================================================================================
- Known Warnings              10
- New Warnings                2
- Known Errors                0
- New Errors                  0
- Panics                      0
================================================================================
- block hash does not match                  10295
- wrong nonce in block                       4061
- miniblocks does not match                  0
- num miniblocks does not match              0
- miniblock hash does not match              0
- block bodies does not match                0
- receipts hash missmatch                    0
================================================================================

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Pull request merged**